_posts/2022-09-07-redis-api-intro.adoc

```diff
@@ -0,0 +1,226 @@
+
+== The new Quarkus Redis API
+
+The entry point of the new Quarkus Redis API are the two _data sources_ interfaces:
+
+* `io.quarkus.redis.datasource.RedisDataSource` - the imperative (blocking) API
+* `io.quarkus.redis.datasource.ReactiveRedisDataSource` - the reactive API
+
+As mentioned above, these APIs are implemented on top of a lower-level client:
+
+image::/assets/images/posts/redis-intro/redis-clients.png[align="center"]
+
+The data source APIs follow the command group structure.
+For each group, you retrieve an object dedicated to emitting the commands of that group.
+In that regard, this new API is not an abstraction of Redis.
+You still need to know the command you need.
+
+For example, to manipulate a `Set<Person>`, you will use the following code:
+
+[source,java]
+----
+record Person(String firstName, String lastName) {}
+
+@ApplicationScoped
+class PersonService {
+
+    private final SetCommands<String, Person> commands;
+
+    public PersonService(RedisDataSource ds) {
+	    // Retrieve the `set` group
+        commands = ds.set(Person.class);
+    }
+
+    public void add(Person person) {
+	    // Emit the `sadd` command
+        commands.sadd("key", person);
+    }
+}
+----
+
+The API manages the serialization and deserialization for you.
+Currently, it uses JSON for objects (using Jackson), but soon the API will offer more advanced features.
+
+This example uses the imperative API, but the reactive API is symmetric.
+
+== Implementing a Redis cache
+
+Time for more cohesive code.
+Let's imagine the following https://github.com/cescoffier/quarkus-redis-demo[application]:
+
+image::/assets/images/posts/redis-intro/application.png[align="center"]
+
+We have a database storing `Heroes`, a lot of them.
+You need to return the 3 most powerful heroes based on their level.
+Of course, you can use your SQL ninja skills, but let's imagine that this code was coded a long time ago, cannot be changed, and is very time-consuming:
+
+[source,java]
+----
+// Dumb approach, don't do this
+return new Ranking(Hero.<Hero>listAll()
+	.stream()
+	.sorted((o1, o2) -> Integer.compare(o2.level, o1.level))
+	.peek(h -> {
+		// do something very long...
+		nap();
+	})
+	.limit(3)
+	.collect(Collectors.toList()));
+----
+
+So, one solution to avoid having to re-compute this set of heroes on every call is to cache the result for a bit of time, let's say 10 seconds.
+Let's consider that it's acceptable in this case to return a potentially outdated result set.
+
+To use the new Redis API, we need to use the `redis-client` extension.
+For users from the previous API, it's the same extension. The previous API is still available but is deprecated, and we plan to remove it at some point.
+
+Now that we can use the  `RedisDataSource`, we can implement the https://github.com/cescoffier/quarkus-redis-demo/blob/main/src/main/java/me/escoffier/quarkus/supes/MyRedisCache.java[`MyRedisCache` class] as follows:
```

Comment:
maybe link to a commit https://github.com/cescoffier/quarkus-redis-demo/blob/735f8cdbe579ef5c13631e8ec6e3ddad6821a3cb/src/main/java/me/escoffier/quarkus/supes/MyRedisCache.java in case the file gets removed/renamed in the new commits on that repo?